File: backend/app/schemas/identidade.py
"""Schemas — Conta, Membro, Papel, Estabelecimento, Convite."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, StringConstraints

from app.models.enums import (
    ContaStatus,
    ContaTipo,
    ConviteStatus,
    PapelInternoMembro,
    PapelStatus,
    PapelTipo,
)
from app.schemas.common import GeoPoint, ORMModel, TimestampedORM


# === Conta ===

class ContaCreate(BaseModel):
    tipo: ContaTipo
    nome_publico: Annotated[str, StringConstraints(min_length=2, max_length=255)]
    cnpj: Annotated[str, StringConstraints(pattern=r"^\d{14}$")] | None = None
    escopo_territorial: dict | None = None


class ContaUpdate(BaseModel):
    nome_publico: str | None = None
    foto_perfil_path: str | None = None
    escopo_territorial: dict | None = None


class ContaRead(TimestampedORM):
    tipo: ContaTipo
    status: ContaStatus
    nome_publico: str
    cnpj: str | None
    foto_perfil_path: str | None
    escopo_territorial: dict | None
    cortesia_ativa: bool


class ContaStatusUpdate(BaseModel):
    """Backoffice — aprova, suspende, anonimiza."""

    status: ContaStatus
    motivo: Annotated[str, StringConstraints(min_length=3, max_length=500)]


# === Membro / Convite ===

class MembroRead(TimestampedORM):
    usuario_id: UUID
    conta_id: UUID
    papel_interno: PapelInternoMembro


class ConviteCreate(BaseModel):
    email: EmailStr
    papel_interno: PapelInternoMembro = PapelInternoMembro.OPERADOR


class ConviteRead(TimestampedORM):
    conta_id: UUID
    email: EmailStr
    papel_interno: PapelInternoMembro
    expira_em: datetime
    status: ConviteStatus


class ConviteAceitar(BaseModel):
    token: Annotated[str, StringConstraints(min_length=20, max_length=128)]


# === Papel Ativado ===

class PapelCreate(BaseModel):
    papel: PapelTipo
    dados_complementares: dict = Field(default_factory=dict)


class PapelUpdate(BaseModel):
    dados_complementares: dict | None = None
    status: PapelStatus | None = None


class PapelRead(TimestampedORM):
    conta_id: UUID
    papel: PapelTipo
    status: PapelStatus
    dados_complementares: dict


# === Estabelecimento ===

class EstabelecimentoCreate(BaseModel):
    nome: Annotated[str, StringConstraints(min_length=2, max_length=255)]
    cep: Annotated[str, StringConstraints(pattern=r"^\d{8}$")]
    logradouro: Annotated[str, StringConstraints(min_length=2, max_length=255)]
    numero: Annotated[str, StringConstraints(max_length=20)]
    complemento: str | None = None
    bairro: Annotated[str, StringConstraints(max_length=100)]
    cidade: Annotated[str, StringConstraints(max_length=100)]
    uf: Annotated[str, StringConstraints(pattern=r"^[A-Z]{2}$")]
    ibge_municipio: Annotated[str, StringConstraints(pattern=r"^\d{7}$")] | None = None
    localizacao: GeoPoint


class EstabelecimentoRead(TimestampedORM):
    conta_id: UUID
    nome: str
    cep: str
    logradouro: str
    numero: str
    complemento: str | None
    bairro: str
    cidade: str
    uf: str
    ibge_municipio: str | None
    lat: float
    lng: float


# === Reputação ===

class ReputacaoPorPapel(ORMModel):
    papel: PapelTipo
    media: float = Field(ge=0, le=5)
    total_avaliacoes: int = Field(ge=0)


class ReputacaoConta(ORMModel):
    conta_id: UUID
    por_papel: list[ReputacaoPorPapel]

File: frontend/src/api/endpoints/identidade.ts
import { api } from '@/api/client';
import type {
  Conta,
  ContaTipo,
  Estabelecimento,
  ID,
  Membro,
  Papel,
  PapelInternoMembro,
  PapelStatus,
  PapelTipo,
} from '@/types/api';

export const contasApi = {
  list: () => api.get<Conta[]>('/contas/').then((r) => r.data),
  get: (id: ID) => api.get<Conta>(`/contas/${id}`).then((r) => r.data),
  create: (p: {
    tipo: ContaTipo;
    nome_publico: string;
    cnpj?: string;
    escopo_territorial?: Record<string, unknown>;
  }) => api.post<Conta>('/contas/', p).then((r) => r.data),
  update: (id: ID, p: Partial<Conta>) => api.patch<Conta>(`/contas/${id}`, p).then((r) => r.data),
};

export const membrosApi = {
  list: (contaId: ID) =>
    api.get<Membro[]>(`/contas/${contaId}/membros/`).then((r) => r.data),
  convidar: (contaId: ID, email: string, papel_interno: PapelInternoMembro = 'operador') =>
    api
      .post(`/contas/${contaId}/membros/`, { email, papel_interno })
      .then((r) => r.data),
  remover: (contaId: ID, membroId: ID) =>
    api.delete(`/contas/${contaId}/membros/${membroId}`).then((r) => r.data),
};

export const papeisApi = {
  list: (contaId: ID) => api.get<Papel[]>(`/contas/${contaId}/papeis/`).then((r) => r.data),
  ativar: (contaId: ID, papel: PapelTipo, dados_complementares: Record<string, unknown> = {}) =>
    api
      .post<Papel>(`/contas/${contaId}/papeis/`, { papel, dados_complementares })
      .then((r) => r.data),
  atualizar: (
    contaId: ID,
    papelId: ID,
    payload: { dados_complementares?: Record<string, unknown>; status?: PapelStatus },
  ) => api.patch<Papel>(`/contas/${contaId}/papeis/${papelId}`, payload).then((r) => r.data),
};

export const estabelecimentosApi = {
  list: (contaId: ID) =>
    api.get<Estabelecimento[]>(`/contas/${contaId}/estabelecimentos/`).then((r) => r.data),
  criar: (
    contaId: ID,
    payload: {
      nome: string;
      cep: string;
      logradouro: string;
      numero: string;
      complemento?: string;
      bairro: string;
      cidade: string;
      uf: string;
      ibge_municipio?: string;
      localizacao: { lat: number; lng: number };
    },
  ) =>
    api
      .post<Estabelecimento>(`/contas/${contaId}/estabelecimentos/`, payload)
      .then((r) => r.data),
};

export const usuariosApi = {
  me: () => api.get('/usuarios/me').then((r) => r.data),
  updateMe: (payload: { nome_completo?: string; telefone?: string; foto_path?: string }) =>
    api.patch('/usuarios/me', payload).then((r) => r.data),
  pedirExclusao: () => api.post('/usuarios/me/excluir').then((r) => r.data),
};
